apps/emqx/src/emqx_packet.erl

```diff
@@ -453,7 +453,7 @@ format_variable(undefined, _) ->
 format_variable(Variable, undefined) ->
     format_variable(Variable);
 format_variable(Variable, Payload) ->
-    io_lib:format("~ts, Payload=~0p", [format_variable(Variable), Payload]).
+    io_lib:format("~ts, Payload=~ts", [format_variable(Variable), Payload]).
```

Comment:
this will be garbled output if not valid string?

Reply:
```erlang
(emqx@127.0.0.1)60>io:format("~0p",[jiffy:encode(#{<<"test">> => 1, <<"is_ok">> => true})]).
<<"{\"test\":1,\"is_ok\":true}">>ok
```
If not vaild string, it will output as 
```erlang
io:format("~0p~n",[<<1,2,3,97,98,99,100,102,250,251,252,254>>]).
<<1,2,3,97,98,99,100,102,250,251,252,254>>
ok
(emqx@127.0.0.1)129> io:format("~ts~n",[<<1,2,3,97,98,99,100,102,250,251,252,254>>]).
^A^B^Cabcdfúûüþ
ok
```

Not valid string means is a User-defined binary protocol (bit_string), we should output as HEX not erlang binary `<<>>`:

```erlang
Bin = binary:encode_hex(<<1,2,3,97,98,99,100,102,250,251,252,254>>).
<<"0102036162636466FAFBFCFE">>
> io:format("~ts~n",[Bin]).
0102036162636466FAFBFCFE
```
We don't have a good way to determine whether it's text or a custom binary protocol. How about add a configuration to let the user decide, defaulting to text?
```bash
trace.payload {
     enable = true,
     output = hex | text
}
```
**FIY**:
```erlang
(emqx@127.0.0.1)61> io:format("~ts",[jiffy:encode(#{<<"test">> => 1, <<"is_ok">> => true})]).
{"test":1,"is_ok":true}ok
(emqx@127.0.0.1)62> io:format("~ts",[jiffy:encode(#{<<"test">> => 1, <<"is_ok">> => true, <<"utf8">> => <<"emqx最棒"/utf8>> })]).
{"utf8":"emqx最棒","test":1,"is_ok":true}ok
(emqx@127.0.0.1)63> io:format("~0p",[jiffy:encode(#{<<"test">> => 1, <<"is_ok">> => true, <<"utf8">> => <<"emqx最棒"/utf8>> })]).
<<123,34,117,116,102,56,34,58,34,101,109,113,120,230,156,128,230,163,146,34,44,34,116,101,115,116,34,58,49,44,34,105,115,95,111,107,34,58,116,114,117,101,125>>
```

Reply:
maybe check at runtime:
https://ninenines.eu/articles/erlang-validate-utf8/

Reply:
This code will return 0 when  `validate_utf8(<<1,2,3>>, 0)` or `validate_utf8(<<"1,2,3">>, 0)`.

Not really what we need:

```erlang
validate_utf8(<<>>, State) -> State;
validate_utf8(<< C, Rest/bits >>, 0) when C < 128 -> validate_utf8(Rest, 0);
```
After a long search on the Internet, I couldn't find a distinction between a custom binary protocol(bit_string) and a string.
If there is a way to do it, it should be `O(n)`, and when the payload is large, it will become a bottleneck. Might as just let the user configure it.
```bash
// hex Hexadecimal  format: 0102030405060708090A0B0C0D0E0F10
// text unicode string format:
// null don't output payload.
trace.payload.output = hex |text |null
 
```
> maybe check at runtime: https://ninenines.eu/articles/erlang-validate-utf8/

Reply:
Okay. Valid utf8 doesn’t always printable.